appserver/handle_topology.go

```diff
@@ -0,0 +1,530 @@
+
+			// Listen to each resource watcher.
+			for w, rwNode := range resourceWatchers {
+				go func(rwNode nodeMeta, w *watcher.Watch) {
+					w.ListenWatcher(func(resourceEvent watch.Event) {
+
+						// Get the resource.
+						r := getResource(resourceEvent.Object)
+						if resourceEvent.Type == "DELETED" {
+
+							// If the event  type was "deleted" delete the resource.
+							nMap.deleteNodeResource(rwNode, r)
+						} else {
+							// If the event was to add create and add it.
+							item := nMap.nodes[rwNode.Name].nd
+							if item.ID == "" {
+								var resource []topology.Resource
+								resource = append(resource, getResource(rwNode.Value))
+								item = topology.NodeData{
+									Resources: resource,
+									ID:        rwNode.ID,
+									Type:      rwNode.Type,
+									Data: topology.Data{
+										URL:          "dummy_url",
+										EditURL:      "dummy_edit_url",
+										BuilderImage: rwNode.Name,
+										DonutStatus:  make(map[string]string),
+									},
+								}
+								var iData innerData
+								iData.nm = rwNode
+								iData.nd = item
+								nMap.nodes[rwNode.Name] = iData
+							}
+							// If the resource does not exist yet, add it. Otherwise,
+							// update the old resource with the new one.
+							nMap.addOrUpdateNodeResource(rwNode.Name, r)
+						}
+
+						// Write the topology.
+						mutex.Lock()
+						ws.WriteJSON(topology.GetSampleTopology(nMap.getNode(), nMap.getResources(), nMap.getGroups(), nMap.getEdges()))
+						mutex.Unlock()
+					})
+				}(rwNode, w)
+			}
+		})
+	}()
+}
+
+// Get topology resources.
+func (nMap nodesMap) getResources() map[string]string {
+	resourceMap := make(map[string]string)
+	for _, iData := range nMap.nodes {
+		if iData.nd.ID != "" {
+			nData, err := json.Marshal(iData.nd)
+			if err != nil {
+				k8log.Error(err, "failed to retrieve json encoding of node")
+			}
+			resourceMap[iData.nd.ID] = string(nData)
+		}
+	}
+
+	return resourceMap
+}
+
+// Get topology edges.
+func (nMap nodesMap) getEdges() []string {
+	var edges []string
+	sourceObjects := make(map[string][]nodeMeta)
+	targetObjects := make(map[string][]string)
+
+	// Arrange keys and target objects.
+	targetObjects = nMap.getAnnotationData("app.openshift.io/connects-to")
+
+	// Arrange keys and source objects.
+	for targetKey, _ := range targetObjects {
+		sourceObjects[targetKey] = append(sourceObjects[targetKey], nMap.getLabelData("app.kubernetes.io/name", targetKey)[targetKey]...)
+	}
+
+	// Lookup the target key in the source key and
+	// construct the edge.
+	for targetKey, targets := range targetObjects {
+		sourceObjects := sourceObjects[targetKey]
+
+		for _, target := range targets {
+			for _, source := range sourceObjects {
+				e, err := json.Marshal(topology.Edge{Source: source.ID, Target: target})
+				if err != nil {
+					k8log.Error(err, "failed to retrieve json encoding of node")
+				}
+				edges = append(edges, string(e))
+			}
+		}
+	}
+
+	return edges
```

Comment:
if no edges are found, edges value must default to an empty array, right now it returns `null`

Reply:
fixed here: bcae39d9879347729a464d0c27a0e0c75a96e019